File: programmers/ForKaKAO/2021/level2/메뉴 리뉴얼.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iostream>

using namespace std;

map<string, int> m[11];
int maxCnt[11] = { 0, };


void sol(string str, int pos, string candi) {
    if (pos >= str.length()) {
        int len = candi.length();
        if (len >= 2) {
            m[len][candi]++;
            maxCnt[len] = max(maxCnt[len], m[len][candi]);
        }
        return;
    }
    sol(str, pos + 1, candi);
    sol(str, pos + 1, candi + str[pos]);
}


vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    
    for (string str : orders) {
        sort(str.begin(), str.end());
        sol(str, 0, "");
    }
    for (int len : course) {
        for (auto item : m[len]) {
            if (item.second >= 2 && item.second == maxCnt[len]) {
                answer.push_back(item.first);
            }
        }
    }
    sort(answer.begin(), answer.end());
    return answer;
}

File: programmers/ForKaKAO/2021/level2/메뉴 리뉴얼.py
import itertools

def solution(orders, course):
    answer = []
    menuMap = [{} for _ in range(11)]
    maxCnt = [0 for _ in range(11)]
    
    for menu in orders:
        for num in range(2, len(menu) + 1):
            for c in itertools.combinations(sorted(menu), num):
                candidate = ''.join(c)
                if candidate not in menuMap[num]:
                    menuMap[num][candidate] = 1
                else:
                    menuMap[num][candidate] += 1
                    maxCnt[num] = max(maxCnt[num], menuMap[num][candidate])
    
    for c in course:
        for k, v in menuMap[c].items():
            if v >= 2 and v == maxCnt[c]:
                answer.append(k)
    answer.sort()
    return answer